src/Razor/src/Microsoft.VisualStudio.RazorExtension/Snippets/SnippetService.cs

```diff
@@ -0,0 +1,221 @@
+using System.Collections.Immutable;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Editor.Razor;
+using Microsoft.VisualStudio.Editor.Razor.Snippets;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.RazorExtension.Snippets;
+
+internal class SnippetService
+{
+    private readonly JoinableTaskFactory _joinableTaskFactory;
+    private readonly IAsyncServiceProvider _serviceProvider;
+    private readonly SnippetCache _snippetCache;
+    private IVsExpansionManager? _vsExpansionManager;
+
+    private readonly object _cacheGuard = new();
+
+    private static readonly Guid s_CSharpLanguageId = new("694dd9b6-b865-4c5b-ad85-86356e9c88dc");
+    private static readonly Guid s_HtmlLanguageId = new("9bbfd173-9770-47dc-b191-651b7ff493cd");
+
+    private static readonly Dictionary<Guid, ImmutableHashSet<string>> s_ignoredSnippets = new()
+    {
+        {
+            // These are identified as the snippets that are provided by the C# Language Server. The list is found
+            // in https://github.com/dotnet/roslyn/blob/8eb40e64564a2a8d44be2fde9b079605f1f10e0f/src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs#L41
+            s_CSharpLanguageId,
+            ImmutableHashSet.Create(
+                "~", "Attribute", "checked", "class", "ctor", "cw", "do", "else", "enum", "equals", "Exception", "for", "foreach", "forr",
+                "if", "indexer", "interface", "invoke", "iterator", "iterindex", "lock", "mbox", "namespace", "#if", "#region", "prop",
+                "propfull", "propg", "sim", "struct", "svm", "switch", "try", "tryf", "unchecked", "unsafe", "using", "while")
+        },
+        {
+            // Currently no HTML snippets are ignored
+            s_HtmlLanguageId, ImmutableHashSet<string>.Empty
+        }
+    };
+
+    public SnippetService(
+            JoinableTaskFactory joinableTaskFactory,
+            IAsyncServiceProvider serviceProvider,
+            SnippetCache snippetCache)
+    {
+        _joinableTaskFactory = joinableTaskFactory;
+        _serviceProvider = serviceProvider;
+        _snippetCache = snippetCache;
+        _ = _joinableTaskFactory.RunAsync(InitializeAsync);
+    }
+
+    private async Task InitializeAsync()
+    {
+        await _joinableTaskFactory.SwitchToMainThreadAsync();
+
+        // Disable VSSDK006 because we use AssumeNotNull instead
+#pragma warning disable VSSDK006 // Check services exist
+        var textManager = (IVsTextManager2?)await _serviceProvider.GetServiceAsync(typeof(SVsTextManager)).ConfigureAwait(false);
+#pragma warning restore VSSDK006 // Check services exist
+
+        textManager.AssumeNotNull();
+        if (textManager.GetExpansionManager(out _vsExpansionManager) == VSConstants.S_OK)
```

Comment:
Wouldn't it be better to check `textManager` for null? The analyzer suppression suggests its possible, and whilst the justification that "we use AssumeNotNull" makes sense for why the analyzer complained, I'm not sure it solves the spirit of the issue. If the service is null then AssumeNotNull will throw, but so will the GetExpansionManager call if the AssumeNotNull call wasn't there. I would have thought the idea was to be resilient to nulls, and not throw.